3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {map} from '../src/utils/object';
+import {dev} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {reportError} from '../src/error';
+initLogConstructor();
+setReportError(reportError);
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * This provides the "glue" between the AMP Analytics tag and the third-party
+ * vendor's metrics-collection page.
+ *
+ * The vendor's page must implement onNewAmpAnalyticsInstance().
+ */
+class AmpAnalytics3pRemoteFrameHelper {
+  constructor(win) {
+    /**
+     * @type {!Array<function(string)>}
+     * @private
+     */
+    this.eventsListeners_ = [];
+
+    /**
+     * @type {!Array<function(string,string)>}
+     * @private
+     */
+    this.extraDataListeners_ = [];
+
+    /**
+     * @type {!Array<Object>}
```

Comment:
Done. I agree Object<string,*> is the best we can do, since it will look like either:
{ senderId: string, type: string, ampAnalytics3pExtraData: object }
or:
{ senderId: string, type: string, ampAnalytics3pEvent: string }